app/src/main/java/com/lightricks/feedexercise/ui/feed/FeedViewModel.kt

```diff
@@ -1,58 +1,69 @@
 package com.lightricks.feedexercise.ui.feed
 
-import androidx.lifecycle.*
-import com.lightricks.feedexercise.data.FeedItem
-import com.lightricks.feedexercise.util.Event
-import java.lang.IllegalArgumentException
+import androidx.lifecycle.ViewModel
+import androidx.lifecycle.ViewModelProvider
+import androidx.lifecycle.viewModelScope
+import com.lightricks.feedexercise.R
+import com.lightricks.feedexercise.domain.FeedError
+import com.lightricks.feedexercise.domain.FeedRedux
+import com.lightricks.feedexercise.domain.FeedRedux.Effect
+import com.lightricks.feedexercise.domain.FeedRedux.Message
+import com.lightricks.feedexercise.domain.FeedRedux.State
+import com.lightricks.feedexercise.ui.feed.model.ViewState
+import com.lightricks.feedexercise.util.mvi.CoroutineEffectHandler
+import com.lightricks.feedexercise.util.mvi.CoroutineMessageEmitter
+import com.lightricks.feedexercise.util.mvi.LaunchReduxEngine
+import com.lightricks.feedexercise.util.mvi.withEffect
+import kotlinx.coroutines.flow.SharingStarted
+import kotlinx.coroutines.flow.StateFlow
+import kotlinx.coroutines.flow.map
+import kotlinx.coroutines.flow.stateIn
+import javax.inject.Inject
 
-/**
- * This view model manages the data for [FeedFragment].
- */
-open class FeedViewModel : ViewModel() {
-    private val stateInternal: MutableLiveData<State> = MutableLiveData<State>(DEFAULT_STATE)
-    private val networkErrorEvent = MutableLiveData<Event<String>>()
-
-    fun getIsLoading(): LiveData<Boolean> {
-        //todo: fix the implementation
-        return MutableLiveData()
-    }
+open class FeedViewModel(
+    feedEffectHandler: CoroutineEffectHandler<Effect, Message>,
+    feedMessageEmitter: CoroutineMessageEmitter<Message>
+) : ViewModel() {
 
-    fun getIsEmpty(): LiveData<Boolean> {
-        //todo: fix the implementation
-        return MutableLiveData()
-    }
-
-    fun getFeedItems(): LiveData<List<FeedItem>> {
-        //todo: fix the implementation
-        return MutableLiveData()
-    }
+    private val engine = LaunchReduxEngine(
+        initial = State.Loading withEffect Effect.GetFeed,
+        reducer = FeedRedux::reduce,
+        messageEmitter = feedMessageEmitter,
+        effectHandler = feedEffectHandler
+    )
 
-    fun getNetworkErrorEvent(): LiveData<Event<String>> = networkErrorEvent
+    val viewState: StateFlow<ViewState> = engine.output
+        .map { state -> createViewState(state) }
+        .stateIn(
+            scope = viewModelScope,
+            started = SharingStarted.WhileSubscribed(5000),
+            initialValue = ViewState.Loading
+        )
```

Comment:
I am not sure I understand why do we need this operator.

Reply:
Mainly it is for unsubscribing when there are no observers - you can read a great post about it (https://medium.com/androiddevelopers/a-safer-way-to-collect-flows-from-android-uis-23080b1f8bda)

Reply:
But here it is a StateFlow (hot flow) so I don't understand the point..

Reply:
Yeah - I know. WhileSubscribed will stop the underlying producer when there are no active observers.
The compose function observes stateFlow. When we call another compose function there is no observer on state flow, so it will stop emitting states after 5000 ms